Source/DafnyStandardLibraries/examples/Actions/ActionsExamples.dfy

```diff
@@ -0,0 +1,177 @@
+module ActionsExamples {
+  import opened Std.Actions
+  import opened Std.Enumerators
+  import opened Std.Aggregators
+  import opened Std.Wrappers
+
+  class Box {
+    const i: nat
+
+    constructor(i: nat)
+      reads {}
+      ensures this.i == i
+    {
+      this.i := i;
+    }
+  }
+
+  function SeqRange(n: nat): seq<nat> {
+    seq(n, i => i)
+  }
+
+  lemma SeqRangeIncr(prefix: seq<nat>, n: nat)
+    requires prefix == SeqRange(n)
+    ensures prefix + [n] == SeqRange(n + 1)
+  {}
+
+  @AssumeCrossModuleTermination
+  class BoxEnumerator extends Action<(), Box> {
+
+    var nextValue: nat
+
+    ghost predicate Valid()
+      reads this, Repr
+      ensures Valid() ==> this in Repr
+      ensures Valid() ==> CanProduce(history)
+      decreases height, 0
+    {
+      && this in Repr
+      && CanProduce(history)
+      && nextValue == |history|
+    }
+
+    constructor()
```

Comment:
Code style: `constructor()` looks like JavaScript to me. I would have written it with a space before the `()`, just like a named constructor `constructor Init()` but with an empty-string name.